File: backend/src/engine/analyzers/entry-points/FrameworkDetector.ts
/**
 * KSA-162: Framework Detector — Identifies frameworks from import statements.
 */

import type { FrameworkInfo } from './types.js';
import { PatternRegistry } from './PatternRegistry.js';

export class FrameworkDetector {
  private registry: PatternRegistry;

  constructor(registry: PatternRegistry) {
    this.registry = registry;
  }

  /** Detect framework from file source code (import analysis). */
  detect(source: string, language: string): FrameworkInfo | null {
    const frameworks = this.registry.getFrameworksForLanguage(language);
    if (frameworks.length === 0) return null;

    let bestMatch: { name: string; score: number } | null = null;

    for (const { name, patterns } of frameworks) {
      let score = 0;
      for (const importPattern of patterns.imports) {
        if (source.includes(importPattern)) {
          score++;
        }
      }
      if (score > 0 && (!bestMatch || score > bestMatch.score)) {
        bestMatch = { name, score };
      }
    }

    if (!bestMatch) return null;

    return {
      name: bestMatch.name,
      language,
      confidence: bestMatch.score >= 2 ? 'High' : 'Medium',
    };
  }

  /** Detect framework from a list of import strings (from relationships table). */
  detectFromImports(imports: string[], language: string): FrameworkInfo | null {
    const frameworks = this.registry.getFrameworksForLanguage(language);
    if (frameworks.length === 0) return null;

    for (const { name, patterns } of frameworks) {
      for (const importPattern of patterns.imports) {
        if (imports.some(imp => imp.includes(importPattern))) {
          return { name, language, confidence: 'High' };
        }
      }
    }
    return null;
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/entry_points/framework_detector.py
"""KSA-162: Framework Detector."""
from __future__ import annotations
from . import FrameworkInfo, Confidence
from .pattern_registry import PatternRegistry


class FrameworkDetector:
    def __init__(self, registry: PatternRegistry) -> None:
        self._registry = registry

    def detect(self, source: str, language: str) -> FrameworkInfo | None:
        frameworks = self._registry.get_frameworks_for_language(language)
        if not frameworks:
            return None
        best: tuple[str, int] | None = None
        for name, patterns in frameworks:
            score = sum(1 for imp in patterns.imports if imp in source)
            if score > 0 and (best is None or score > best[1]):
                best = (name, score)
        if not best:
            return None
        confidence = Confidence.High if best[1] >= 2 else Confidence.Medium
        return FrameworkInfo(best[0], language, confidence)
